learn/security/tenant_tokens.md

```diff
@@ -0,0 +1,299 @@
+
+The full process requires you to [create a token header](#step-1-creating-a-token-header), [prepare a data payload](#step-2-preparing-the-token-payload) with at least one set of search rules, and then [sign the token](#step-3-signing-a-token) with an API key.
+
+If you are already familiar with the creation of JWTs and only want to know about the specific requirements of a tenant token payload, take a look at this guide's [token payload reference](#tenant-token-payload-reference).
+
+::: note
+The following examples use JavaScript for the sake of simplicity, but Meilisearch should be compatible with the majority of modern stacks and languages.
+:::
+
+### Step 1: Creating a token header
+
+The first step when generating a tenant token is to create a header stating token type and the chosen encryption algorithm. Once that is done you must encode the header object into `base64`:
+
+```js
+let header = {
+  "alg": "HS256",
+  "typ": "JWT"
+}
+
+const base64Header = base64Encode(header);
+```
+
+The encryption algorithm indicated by `alg` can be one of: `HS256`, `HS384`, or `HS512`. The choice of algorithm has no impact in tenant token behavior.
+
+Meilisearch only supports JWT tokens and `typ` must always be `JWT`.
+
+### Step 2: Preparing the token payload
+
+Once the header is ready, you must create a payload. In this case, payload means the data containing the token's API key prefix, search rules, and optional expiry date.
+
+First, you must extract the first eight characters from a valid API key with access to the search endpoint:
+
+```js
+const meiliSearchApikey = '[an_api_key]';
+const yourApiKeyPrefix = meiliSearchApiKey.slice(0,8);
+```
+
+`apiKey` must be a valid Meilisearch API key. A tenant token will have access to the same indexes as the API key you use when generating it. When generating a tenant token payload, you should only supply the first 8 characters of a key.
+
+Your payload must also contain a `searchRules` object. The following example would force all searches made with the token to only return results with a `user_id` of `1`:
+
+```js
+const yourSearchRules = {
+  "patient_medical_records": {
+    "filter": "user_id = 1"
+  }
+}
+```
+
+`searchRules` must be an object containing at least one search rule. A search rule must specify either an index name or a `*` wildcard. Search rules must also declare at least one filter. In this example, the filter makes it so a user only has access to their own records.
+
+With both these steps completed, you are ready to generate the token payload:
+
+```js
+const payload = {
+  "apiKeyPrefix": yourApiKeyPrefix,
+  "exp": 1641835850,
+  "searchRules": yourSearchRules
+}
+```
+
+You can also optionally set an expiry date for your token. Tokens are invalid once past their expiry date.
+
+Like the header, it is necessary to encode the payload:
+
+```js
+const base64Payload = base64Encode(payload);
+```
```

Comment:
Same as above:
```
function encode64(str: any) {
  return Buffer.from(JSON.stringify(str)).toString('base64')
}
```

Reply:
the name of my variable is horrible, it should be 
```js
function encore64(data) {
  return Buffer.from(JSON.stringify(data)).toString('base64')
}
```

Then this function `JSON.stringify(data)` hidden in the middle, is transforming any data into a string (if it is a valid JSON but there is no reason why the user would use a non valid one, in which case it will throw an error).